toolchain/check/impl_lookup.cpp

```diff
@@ -566,48 +568,206 @@ static auto GetOrAddLookupImplWitness(Context& context, SemIR::LocId loc_id,
+                                 SemIR::InstId facet_value_inst_id,
+                                 SemIR::InterfaceId& destroy_interface_id)
+    -> bool {
+  if (!facet_value_inst_id.has_value()) {
+    return false;
+  }
+  auto facet_value = context.insts().Get(facet_value_inst_id);
+  auto facet_type =
+      context.types().GetAs<SemIR::FacetType>(facet_value.type_id());
+  return CanDestroyFacetType(context, loc_id, facet_type, destroy_interface_id);
+}
+
+// Returns true if the `Self` should impl `Destroy`.
+static auto TypeCanDestroy(Context& context, SemIR::LocId loc_id,
+                           SemIR::ConstantId query_self_const_id) -> bool {
+  // A lazily cached `Destroy` interface.
+  auto destroy_interface_id = SemIR::InterfaceId::None;
+
+  auto query_inst_id = context.constant_values().GetInstId(
+      GetCanonicalFacetOrTypeValue(context, query_self_const_id));
+
+  // When querying a facet type, we can return early.
+  if (auto facet_type = context.types().TryGetAs<SemIR::FacetType>(
+          context.insts().Get(query_inst_id).type_id())) {
+    return CanDestroyFacetType(context, loc_id, *facet_type,
+                               destroy_interface_id);
+  }
+
+  // A stack of IDs to check (ordering shouldn't be important).
+  struct WorkItem {
+    SemIR::InstId id;
+    // The only case `abstract` is allowed during a type walk is when looking at
+    // a `base` type of a class.
+    bool allow_abstract = false;
+  };
+  llvm::SmallVector<WorkItem> work;
+
+  // Start by checking the queried instruction.
+  work.push_back({.id = query_inst_id});
+
+  // Loop through type structures recursively, looking for any types
+  // incompatible with `Destroy`.
+  while (!work.empty()) {
+    auto work_item = work.pop_back_val();
+    auto inst = context.insts().Get(work_item.id);
+
+    CARBON_KIND_SWITCH(inst) {
+      case CARBON_KIND(SemIR::ArrayType array_type): {
+        work.push_back({.id = array_type.element_type_inst_id});
+        break;
+      }
+
+      case CARBON_KIND(SemIR::BaseDecl base_decl): {
+        work.push_back(
+            {.id = base_decl.base_type_inst_id, .allow_abstract = true});
+        break;
+      }
+
+      case CARBON_KIND(SemIR::ClassType class_type): {
+        auto class_info = context.classes().Get(class_type.class_id);
+        // Incomplete and abstract classes can't be destroyed.
+        if (!class_info.is_complete() ||
+            (!work_item.allow_abstract &&
+             class_info.inheritance_kind ==
+                 SemIR::Class::InheritanceKind::Abstract)) {
+          return false;
+        }
+
+        // For C++ types, use Clang to determine whether they can be destructed.
+        // TODO: Needs appropriate calls.
+        if (context.name_scopes().Get(class_info.scope_id).is_cpp_scope()) {
+          break;
+        }
+
+        auto obj_repr_id =
+            class_info.GetObjectRepr(context.sem_ir(), SemIR::SpecificId::None);
```

Comment:
```suggestion
        auto obj_repr_id =
            class_info.GetObjectRepr(context.sem_ir(), class_type.specific_id);
```
I think this might allow you to undo a bunch of the test changes.

Reply:
This doesn't work right now, adding a TODO